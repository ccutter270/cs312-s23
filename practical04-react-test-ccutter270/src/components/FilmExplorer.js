import React, { useState, useEffect } from "react";

import FilmTableContainer from "./FilmTableContainer";
import SearchBar from "./SearchBar";




function FilmExplorer({rawData}) {
  const [searchTerm, setSearchTerm] = useState("");
  const [sortType, setSortType] = useState("title");
  const [films, setFilms] = useState([]);
  const [ascending, setDirection] = useState(true);

  // load the film data
  useEffect(() => {
    if (rawData){
      setFilms(rawData);
    }
  }, [rawData]);


   // change the rating of a film
   const setRating = (filmid, rating) => {
    const alteredFilms = films.map((film) => {
      if (film.id === filmid) {
        return { ...film, rating };
      }
      return film;
    });
    setFilms(alteredFilms);
  };

  const mainContents =
    films.length === 0 ? (
      <h2>Loading...</h2>
    ) : (
      <FilmTableContainer
        searchTerm={searchTerm}
        films={films}
        sortType={sortType}
        setRatingFor={setRating}
        ascending={ascending}
      />
    );

  return (
    <div className="FilmExplorer">
      <SearchBar
        searchTerm={searchTerm}
        setTerm={setSearchTerm}
        sortType={sortType}
        setType={setSortType}
        ascending={ascending}
        setDirection={setDirection}
      />
      {mainContents}
    </div>
  );
}

export default FilmExplorer;
