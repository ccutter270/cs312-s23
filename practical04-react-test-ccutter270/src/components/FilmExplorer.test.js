import { render, screen, fireEvent } from "@testing-library/react";

import FilmExplorer from "./FilmExplorer";

const films = [
  {
    id: 101,
    overview: "The first Thin Man movie",
    release_date: "1934-05-25",
    poster_path: "/rEXmQIgG8aHiZZbYAPdJjZNUxa.jpg",
    title: "The Thin Man",
    vote_average: 7.7,
  },
  {
    id: 102,
    overview: "The second Thin Man movie",
    release_date: "1936-12-25",
    poster_path: "/1fkCHOffEkPOY0UwSrvzG6YKDVp.jpg",
    title: "After the Thin Man",
    vote_average: 7.6,
  },
];

describe("Testing FilmExplorer", () => {
  let testFilms;

  beforeEach(() => {
    testFilms = films.map((film) => ({ ...film }));
  });

  // TODO: Add your smoke and snapshot tests

  test("Smoke test", () => {
    render(<FilmExplorer />);
  });

  test("Snapshot test", () => {
    const { container } = render(<FilmExplorer />);
    expect(container.firstChild).toMatchSnapshot();
  });

  // TODO: Add your "Arrow changes direction" test

  test("Arrow changes direction", () => {
    // 1. Render Film Explorer
    render(<FilmExplorer rawData={testFilms} />);

    // 2. Check that "▲" is in the document
    let arrow = screen.queryByText("▲");
    expect(arrow).toBeInTheDocument();

    // 3. Simulate a click on the up arrow
    fireEvent.click(arrow);

    // 4. Check that "▲" is no longer in the document
    arrow = screen.queryByText("▲");
    expect(arrow).not.toBeInTheDocument();

    // 5. Check that "▼" is in the document
    arrow = screen.queryByText("▼");
    expect(arrow).toBeInTheDocument();

    // 6. Simulates a click on the down arrow
    fireEvent.click(arrow);

    // 7. Check that "▼" is no longer in the document
    arrow = screen.queryByText("▼");
    expect(arrow).not.toBeInTheDocument();

    // 8. Check that "▲" is in the document1.
    arrow = screen.queryByText("▲");
    expect(arrow).toBeInTheDocument();
  });

  test("Rating changes", () => {
    render(<FilmExplorer rawData={testFilms} />);
    let stars = screen.getAllByText("★");

    expect(stars.filter((star) => star.className === "empty")).toHaveLength(10);
    expect(stars.filter((star) => star.className === "filled")).toHaveLength(0);

    fireEvent.click(stars[4]);

    stars = screen.getAllByText("★");
    expect(stars.filter((star) => star.className === "empty")).toHaveLength(5);
    expect(stars.filter((star) => star.className === "filled")).toHaveLength(5);
  });
});
