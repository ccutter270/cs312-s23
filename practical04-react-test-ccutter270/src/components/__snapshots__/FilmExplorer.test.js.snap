// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing FilmExplorer Snapshot test 1`] = `
<div
  class="FilmExplorer"
>
  <div
    class="container"
  >
    <h1
      class="title"
    >
      Film Explorer!
    </h1>
    <div
      class="contents"
    >
      <input
        placeholder="Search"
        type="text"
        value=""
      />
      <p>
        order by 
        <select
          class="select"
        >
          <option
            value="title"
          >
            Title
          </option>
          <option
            value="release_date"
          >
            Date
          </option>
          <option
            value="vote_average"
          >
            TMDB Rating
          </option>
        </select>
         
        <span
          class="arrow"
        >
          â–²
        </span>
      </p>
    </div>
  </div>
  <h2>
    Loading...
  </h2>
</div>
`;
